File: code/python/lists/linked_queue.py
from no_element import NoElement


class Node:
    def __init__(self, item, nxt=None):
        self.item = item
        self.next = nxt


class LinkedQueue:
    """A queue based on singly linked lists."""

    def __init__(self):
        self._head = None
        self._tail = None
        self._size = 0

    def size(self):
        return self._size

    def is_empty(self):
        return self._head is None

    def peek(self):
        if self.is_empty():
            raise NoElement('Peek from an empty queue')
        return self._head.item

    # enqueue() = add_last()
    def enqueue(self, item):
        old_tail = self._tail
        self._tail = Node(item)
        if self.is_empty():
            self._head = self._tail
        else:
            old_tail.next = self._tail
        self._size += 1

    # dequeue() = remove_first()
    def dequeue(self):
        if self.is_empty():
            raise NoElement('Dequeue from an empty queue')
        answer = self._head.item
        self._head = self._head.next
        if self.is_empty():
            self._tail = None
        self._size -= 1
        return answer

    def __iter__(self):
        return LinkedIterator(self._head)


class LinkedIterator:
    def __init__(self, head: Node):
        self.walk = head

    def __next__(self):
        if self.walk is None:
            raise StopIteration
        answer = self.walk.item
        self.walk = self.walk.next
        return answer

File: code/java/lists/src/main/java/org/swufe/datastructure/LinkedQueue.java
package org.swufe.datastructure;

import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * A queue based on singly linked lists.
 */
public class LinkedQueue<Item> implements Iterable<Item> {
    private static class Node<Item> {
        Item item;
        Node<Item> next;

        Node(Item item) {
            this.item = item;
            next = null;
        }
    }

    private Node<Item> head;
    private Node<Item> tail;
    private int size;

    public LinkedQueue() {
        head = null;
        tail = null;
        size = 0;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return head == null;
    }

    public Item peek() {
        if (isEmpty()) throw new NoSuchElementException("Peek from an empty queue");
        return head.item;
    }

    // enqueue() = add_last()
    public void enqueue(Item item) {
        Node<Item> oldTail = tail;
        tail = new Node<>(item);
        if (isEmpty()) head = tail;
        else oldTail.next = tail;
        size += 1;
    }

    // dequeue() = remove_first()
    public Item dequeue() {
        if (isEmpty()) throw new NoSuchElementException("Peek from an empty queue");
        Item answer = head.item;
        head = head.next;
        if (isEmpty()) {
            tail = null;
        }
        size -= 1;
        return answer;
    }

    @Override
    public Iterator<Item> iterator() {
        return new LinkedIterator();
    }

    private class LinkedIterator implements Iterator<Item> {
        private Node<Item> node = head;

        @Override
        public boolean hasNext() {
            return node != null;
        }

        @Override
        public Item next() {
            Item answer = node.item;
            node = node.next;
            return answer;
        }
    }
}
